cl2/density: add pod-pod anti-affinity benchmarks

adtac opened the pull request:
/cc @ahg-g @mm4tt 
/sig scheduling

(Will update with results in a bit.)

ahg-g commented on clusterloader2/testing/density/suite.yaml:
```diff
@@ -0,0 +1,8 @@
+  configPath: testing/density/config.yaml
+  overridePaths: []
+
+- identifier: pod-pod-anti-affinity
```
lets rename to pod-anti-affinity (drop the first pod), here and in the directory names as well.

adtac replied:
my thinking behind this was that we'll be adding pod-to-node affinity too (similar naming as the integration tests)

but since pod will be a common prefix, I think we can drop it :+1:

ahg-g replied:
I think we can use pod-anti-affinity/pod-affinity and node-anti-affinity/node-affinity.

ahg-g commented on clusterloader2/testing/density/pod-pod-anti-affinity/overrides.yaml:
```diff
@@ -0,0 +1,7 @@
+# The default CL2_SCHEDULER_THROUGHPUT_THRESHOLD value of 90 is too high for
+# pod-pod anti-affinity. Use a low threshold until a good baseline can be
+# identified and then increase the threshold incrementally.
+CL2_SCHEDULER_THROUGHPUT_THRESHOLD: 40
```
is this for an alert? if so, please clarify that in the comment and I think we should keep it at 90 unless this will be executed for older versions.

adtac replied:
oops, this is an artifact from my testing when the throughput was below 50 for unrelated reasons. removing.

ahg-g commented on clusterloader2/testing/density/suite.yaml:
```diff
@@ -0,0 +1,8 @@
+- identifier: vanilla
```
out of curiosity, where is this suite.yaml file being consumed? and how do we ensure that this only gets executed on kubemark?

adtac replied:
It's not used here, but we'll need to update kubernetes/test-infra's scalability tests to use the suite instead of directly running the config.yaml

ahg-g commented:
/ok-to-test

adtac commented on clusterloader2/testing/density/config.yaml:
```diff
@@ -15,7 +15,7 @@
 {{$LATENCY_POD_CPU := DefaultParam .LATENCY_POD_CPU 100}}
 {{$LATENCY_POD_MEMORY := DefaultParam .LATENCY_POD_MEMORY 350}}
-{{$MIN_LATENCY_PODS := 500}}
+{{$MIN_LATENCY_PODS := MinInt (MultiplyInt $PODS_PER_NODE .Nodes) 500}}
```
`PODS_PER_NODE * Nodes` will be greater than 500 in all cases that we currently run; existing tests are therefore unaffected. However, in the requiredDuringScheduling anti-affinity tests, we cannot schedule more than one pod per node, so we need this.

mm4tt commented:
> /hold
> 
> investigating an issue in my local tests, will report back

SG, please ping when the PR is ready to review

adtac commented:
Somewhere along the process, I changed `podAffinityTerm` to `podAntiAffinityTerm`, which does not exist even in pod anti-affinity. Anyway, I've fixed that and I've re-run the experiments locally and everything looks good. Here are the results:

```
~/tmp/cl2/032 $ cat SchedulingThroughput_density_2020-08-24T19:47:33Z.json                                                                                                                                                                                                                                                                          
{                                                                                                                                                                                                                                                                                                                                                                            
  "perc50": 66.6,                                                                                                                                                                                                                                                                                                                                                            
  "perc90": 103.8,                                                                                                                                                                                                                                                                                                                                                           
  "perc99": 103.8,                                                                                                                                                                                                                                                                                                                                                           
  "max": 103.8                                                                             
}                                                                                          
```

Interestingly, there's a considerable drop in throughput when scheduling 3000 pods with preferred anti-affinity on 100 nodes. The ~66.6 pods/sec is reproducible across multiple experiments (in the 65-72 range across 5 runs), so it's definitely not measurement error or noise.

/hold cancel

adtac commented:
/cc @mm4tt

ahg-g commented:
/lgtm

mm4tt commented on clusterloader2/testing/density/suite.yaml:
```diff
@@ -0,0 +1,8 @@
+- identifier: vanilla
+  configPath: testing/density/config.yaml
+  overridePaths: []
```
Let's create a separate suite just for scheduler benchmarks. We'll run the vanilla as usual and start running the scheduler suite once a day or so.

adtac replied:
Done.

ahg-g commented:
> Somewhere along the process, I changed `podAffinityTerm` to `podAntiAffinityTerm`, which does not exist even in pod anti-affinity. Anyway, I've fixed that and I've re-run the experiments locally and everything looks good. Here are the results:
> 
> ```
> ~/tmp/cl2/032 $ cat SchedulingThroughput_density_2020-08-24T19:47:33Z.json                                                                                                                                                                                                                                                                          
> {                                                                                                                                                                                                                                                                                                                                                                            
>   "perc50": 66.6,                                                                                                                                                                                                                                                                                                                                                            
>   "perc90": 103.8,                                                                                                                                                                                                                                                                                                                                                           
>   "perc99": 103.8,                                                                                                                                                                                                                                                                                                                                                           
>   "max": 103.8                                                                             
> }                                                                                          
> ~/tmp/cl2/033 $ cat SchedulingThroughput_density_2020-08-24T19:55:56Z.json 
> {
>   "perc50": 71.2,
>   "perc90": 100.2,
>   "perc99": 100.2,
>   "max": 100.2
> }
> ```
> 
> Interestingly, there's a considerable drop in throughput when scheduling 3000 pods with preferred anti-affinity on 100 nodes. The ~66.6 pods/sec is reproducible across multiple experiments (in the 65-72 range across 5 runs), so it's definitely not measurement error or noise.
> 
> /hold cancel

Thanks Adi, I wonder what master size was used?

adtac commented:
>Thanks Adi, I wonder what master size was used?

The kubemark master was a n1-standard-4 (4 vCPUs, 16 GB memory). Note that this was the same master used in vanilla tests, which produced a ~100 pods/sec throughput.

mm4tt commented on clusterloader2/testing/density/suite.yaml:
```diff
@@ -0,0 +1,4 @@
+- identifier: pod-anti-affinity
```
Can we name the file `scheduler-suite.yaml` or something like this?

Other than that it looks good to me

/approve

adtac replied:
Done.

ahg-g commented:
/lgtm
